### PR TITLE
Task vv/117 add shownotes for episode functionality

### PR DESCRIPTION
**Pull request by vitalik-prog**
image

### CODE REVIEW

File: frontend/src/components/episode/components/shownotes-list/shownotes-list.tsx [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+};
+
+const ShownotesList: React.FC<Props> = ({ shownotes, handleJumpToTimeLine }) => {
+  const halfOfShownotes = shownotes.length / 2;
+  const notesOnLeftSide = [];
+  const notesOnRightSide = [];
+
+  for (let i = 0; i < shownotes.length; i++) {
+    i < halfOfShownotes ? notesOnLeftSide.push(shownotes[i]) : notesOnRightSide.push(shownotes[i]);
+  }
```
Comment (ghost) [inline frontend/src/components/episode/components/shownotes-list/shownotes-list.tsx]: do it with css

File: frontend/src/components/episode/components/shownotes/shownotes.tsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+const Shownotes: React.FC<Props> = ({ shownotes, handleJumpToTimeLine }) => {
+
+  return (
+    <ul className={styles.list}>
+      {shownotes.map((shownote) => (
+        <li
+          key={shownote.id}
+          className={styles.rowWrapper}
+          onClick={(): void => handleJumpToTimeLine(shownote.timestamp)}
```
Comment (ghost) [inline frontend/src/components/episode/components/shownotes/shownotes.tsx]: remove anonymous function from jsx

Comment (ghost) [inline frontend/src/components/episode/components/shownotes/shownotes.tsx]: remove `onClick` from `li`

File: frontend/src/components/episode/episode.tsx [hunk 3]
```diff
@@ -36,6 +37,10 @@ const Episode: React.FC = () => {
     dispatch(episodeActions.loadEpisode(Number(id)));
   }, []);
 
+  const handleJumpToTimeLine = (timeline:number): void => {
```
Comment (ZubkoWhatislove) [inline frontend/src/components/episode/episode.tsx:40]:
```suggestion
  const handleJumpToTimeline = (timeline:number): void => {
```

File: frontend/src/components/episode/episode.tsx [hunk 4]
```diff
@@ -60,6 +65,12 @@ const Episode: React.FC = () => {
               <p className={styles.description}>{episode.description}</p>
               <p className={styles.type}>Type: {episode.type}</p>
               <p className={styles.type}>Status: {episode.status}</p>
+              {episode.shownotes && (
```
Comment (ghost) [inline frontend/src/components/episode/episode.tsx]: array is always true

File: frontend/src/components/episode/helpers/shownote-time/shownote-time.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import { MILLISECONDS_IN_SECOND } from '../../../common/player/common/constants';
```
Comment (ghost) [inline frontend/src/components/episode/helpers/shownote-time/shownote-time.ts]: remove it and try to use `getFormattedDate` helper

File: shared/src/common/types/episode/episode.type.ts [hunk 6]
```diff
@@ -9,6 +10,7 @@ type Episode = {
   imageId: number | null;
   image: Image | null;
   record: Record | null;
+  shownotes: Shownote[] | null;
```
Comment (ghost) [inline shared/src/common/types/episode/episode.type.ts]:
```suggestion
  shownotes: Shownote[];
```

**Review by ghost — CHANGES_REQUESTED** [review]

File: frontend/src/components/episode/episode.tsx [hunk 7]
```diff
@@ -62,6 +66,12 @@ const Episode: React.FC = () => {
               <p className={styles.description}>{episode.description}</p>
               <p className={styles.type}>Type: {episode.type}</p>
               <p className={styles.type}>Status: {episode.status}</p>
+              {episode.shownotes?.length ? (
+                <div className={styles.shownotesWrapper}>
+                  <h3>Time navigation</h3>
+                  <ShownotesList shownotes={episode.shownotes} handleJumpToTimeLine={handleJumpToTimeLine} />
+                </div>
+              ) : null}
```
Comment (ghost) [inline frontend/src/components/episode/episode.tsx]:
```suggestion
              {episode.shownotes.length && (
                <div className={styles.shownotesWrapper}>
                  <h3>Time navigation</h3>
                  <ShownotesList shownotes={episode.shownotes} handleJumpToTimeLine={handleJumpToTimeLine} />
                </div>
              )}
```

File: frontend/src/components/common/common.ts [hunk 8]
```diff
@@ -21,3 +21,5 @@ export { default as PodcastList } from './podcasts-list/podcasts-list';
 export { default as PodcastItem } from './podcast-item/podcast-item';
 export { default as Player } from './player/player';
 export { default as DefaultImage } from './default-image/default-image';
+export { default as ShownotesList } from './shownotes-list/shownotes-list';
+export { default as Shownote } from './shownote/shownote';
```
Comment (ghost) [inline frontend/src/components/common/common.ts]: these components are specific
move it to page where we use it

File: frontend/src/components/common/shownote/shownote.tsx [hunk 9]
```diff
@@ -0,0 +1,37 @@
+import { Shownote as TShownote } from 'common/types/types';
+import { DateFormatType } from 'common/enums/enums';
+import styles from './styles.module.scss';
+import { getFormattedDate } from 'helpers/date/date';
+import { MILLISECONDS_IN_SECOND } from '../player/common/constants';
```
Comment (ghost) [inline frontend/src/components/common/shownote/shownote.tsx]: in this case move it to the `frontend/src/common/constants/`

File: frontend/src/components/common/shownote/shownote.tsx [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+type Props = {
+  shownote: TShownote;
+  handleJumpToTimeLine: (payload: number) => void;
+};
+
+const Shownote: React.FC<Props> = ({ shownote, handleJumpToTimeLine }) => {
+  const timeZoneOffsetInMS = new Date().getTimezoneOffset() * 60 * MILLISECONDS_IN_SECOND;
+  const offset = new Date(shownote.timestamp * 1000 + timeZoneOffsetInMS);
+  const time = getFormattedDate(offset, DateFormatType.HOURS_MINUTES_SECONDS);
```
Comment (ghost) [inline frontend/src/components/common/shownote/shownote.tsx]: collect it into one helper and move it to `frontend/src/components/common/shownote/helpers/`

File: frontend/src/components/common/shownote/shownote.tsx [hunk 11]
```diff
@@ -0,0 +1,37 @@
+import { Shownote as TShownote } from 'common/types/types';
+import { DateFormatType } from 'common/enums/enums';
+import styles from './styles.module.scss';
+import { getFormattedDate } from 'helpers/date/date';
+import { MILLISECONDS_IN_SECOND } from '../player/common/constants';
+
+type Props = {
+  shownote: TShownote;
+  handleJumpToTimeLine: (payload: number) => void;
```
Comment (ghost) [inline frontend/src/components/common/shownote/shownote.tsx]: check **A8**

File: frontend/src/components/common/shownote/shownote.tsx [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+};
+
+const Shownote: React.FC<Props> = ({ shownote, handleJumpToTimeLine }) => {
+  const timeZoneOffsetInMS = new Date().getTimezoneOffset() * 60 * MILLISECONDS_IN_SECOND;
+  const offset = new Date(shownote.timestamp * 1000 + timeZoneOffsetInMS);
+  const time = getFormattedDate(offset, DateFormatType.HOURS_MINUTES_SECONDS);
+
+  const onTimeLineJump = (): void => {
```
Comment (ghost) [inline frontend/src/components/common/shownote/shownote.tsx]: check **A10**

File: frontend/src/components/common/shownote/shownote.tsx [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+      onClick={onTimeLineJump}
+    >
+      <span className={styles.time}>
+        {time}
+      </span>
+      &nbsp;
+      <span className={styles.noteName}>
+        {shownote.name}
+      </span>
+    </button>
```
Comment (ghost) [inline frontend/src/components/common/shownote/shownote.tsx]:
```suggestion
<li>
    <button
      className={styles.rowWrapper}
      onClick={onTimeLineJump}
    >
      <span className={styles.time}>
        {time}
      </span>
      &nbsp;
      <span className={styles.noteName}>
        {shownote.name}
      </span>
    </button>
</li>
```

File: frontend/src/components/common/shownotes-list/shownotes-list.tsx [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+      className={styles.columnsWrapper}
+    >
+      {shownotes.map((shownote) => (
+        <Shownote
+          key={shownote.id}
+          shownote={shownote}
+          handleJumpToTimeLine={handleJumpToTimeLine}
+        />
+      ))}
+    </div>
```
Comment (ghost) [inline frontend/src/components/common/shownotes-list/shownotes-list.tsx]:
```suggestion
    <ul
      style={getGridMarkup(shownotes)}
      className={styles.columnsWrapper}
    >
      {shownotes.map((shownote) => (
        <Shownote
          key={shownote.id}
          shownote={shownote}
          handleJumpToTimeLine={handleJumpToTimeLine}
        />
      ))}
    </ul>
```

**Review by ghost — CHANGES_REQUESTED** [review]

File: frontend/src/components/common/shownotes-list/shownotes-list.tsx [hunk 15]
```diff
@@ -0,0 +1,47 @@
+import { Shownote as TShownote } from 'common/types/types';
+import { Shownote } from '../common';
+import styles from './styles.module.scss';
+
+type gridStyle = {
+  gridTemplateColumns: string;
+  gridTemplateRows: string;
+};
+
+const getGridMarkup = (shownotes: TShownote[]): gridStyle => {
```
Comment (ghost) [inline frontend/src/components/common/shownotes-list/shownotes-list.tsx]: just to it in styles 🙂

**Review by ghost — CHANGES_REQUESTED** [review]

File: frontend/src/components/episode/helpers/get-time-offset/get-time-offset.ts [hunk 16]
```diff
@@ -0,0 +1,14 @@
+import { DateFormatType } from 'common/enums/enums';
+import { Shownote as TShownote } from 'common/types/types';
+import { MILLISECONDS_IN_SECOND, SECONDS_IN_MINUTE } from '../../../../common/constants/constants';
```
Comment (ghost) [inline frontend/src/components/episode/helpers/get-time-offset/get-time-offset.ts]:
```suggestion
import { MILLISECONDS_IN_SECOND, SECONDS_IN_MINUTE } from 'common/constants/constants';
```

**Review by ghost — DISMISSED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**